packages/imperative/src/events/src/ImperativeEventEmitter.ts

```diff
@@ -0,0 +1,288 @@
+/*
+* This program and the accompanying materials are made available under the terms of the
+* Eclipse Public License v2.0 which accompanies this distribution, and is available at
+* https://www.eclipse.org/legal/epl-v20.html
+*
+* SPDX-License-Identifier: EPL-2.0
+*
+* Copyright Contributors to the Zowe Project.
+*
+*/
+
+import * as fs from "fs";
+import { homedir } from "os";
+import { join } from "path";
+import { ImperativeConfig } from "../../utilities/src/ImperativeConfig";
+import { ImperativeError } from "../../error/src/ImperativeError";
+import { ImperativeEventType, ImperativeUserEvents, ImperativeSharedEvents } from "./ImperativeEventConstants";
+import { ImperativeEvent } from "./ImperativeEvent";
+import { Logger } from "../../logger/src/Logger";
+import { LoggerManager } from "../../logger/src/LoggerManager";
+import { IImperativeRegisteredAction, IImperativeEventEmitterOpts, IImperativeEventJson } from "./doc";
+import { ConfigUtils } from "../../config/src/ConfigUtils";
+
+export class ImperativeEventEmitter {
+    private static mInstance: ImperativeEventEmitter;
+    private static initialized = false;
+    private subscriptions: Map<string, [fs.FSWatcher, Function[]]>;
```

Comment:
Should we investigate using a stronger type for the callback functions? Otherwise, the current `Function[]` type will allow a user to provide a function with multiple parameters as a callback, even though the current logic does not provide any arguments to these callbacks when fired.

Something like this would suffice:

```ts
type EventCallback = () => void;
private subscriptions: Map<string, [fs.FSWatcher, EventCallback[]]>;
```

---

Also, just a note for awareness: the current implementation will only support synchronous callbacks. I'm not sure if this is intended, but if you wanted to support both, the callback type above could be adjusted to handle both return types:

```ts
type EventCallback = () => void | PromiseLike<void>;

// later, when firing callbacks...
for (const cb in callbacks) {
    await cb();
}
```

Keep in mind that support for `async` callbacks would also involve updating the function passed to the `fs.watch` call.

Reply:
I believe I can allow the callback to be async, however, I don't think we can guarantee that there won't be race conditions  due to how the Node's WatchListener is defined.

I'll do some testing on this again to see what can be done. 😋

Reply:
Yep you're right, sorry about that - the current implementation will allow firing `async` callbacks, but as you mentioned the order cannot be guaranteed without `await`ing each callback.

If that's the case and we are okay with the implications of that, then you can disregard the comment about async. callbacks 😋

Reply:
I'd still want to do some testing to see how it reflects on a working VSCE 😋

Reply:
As we thought, all async callbacks will be fired in order, but no guarantee of who finishes first.
Also, it's possible that new emissions happen before the async callback is finished, in which case the results may be unpredictable 😋